crates/dns/src/proxy.rs

```diff
@@ -0,0 +1,261 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::resolver::{Answer, Resolver};
+use std::sync::Arc;
+use tracing::{error, warn};
+use trust_dns_proto::op::{Edns, Header, MessageType, OpCode, ResponseCode};
+use trust_dns_proto::rr::Record;
+use trust_dns_resolver::error::ResolveErrorKind;
+use trust_dns_server::authority::{LookupError, MessageResponse, MessageResponseBuilder};
+use trust_dns_server::server::{Request, RequestHandler, ResponseHandler, ResponseInfo};
+
+/// A Trust-DNS [ResponseHandler] that proxies all DNS requests.
+///
+/// A proxy is fundamentally different than an `Authority` in TrustDNS, since the answers may
```

Comment:
Do we know if upstream would be amenable to this in theory?

it's hard to grok the cost of carrying this (more-verbose) workaround if we don't have a read on whether it can ever be upstreamed or not.

Reply:
I'd like to land this before I approach the Trust-DNS folks. It definitely seems to be a use case they hadn't considered in it's current form, so I think it's worth at least having the discussion. Even if they don't accept it, it's not a lot of code and is stuff that needs to be done if we're going to use Trust-DNS anyway.